chore: refactor recipient validation in new send implementation

jpuri opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Refactor recipient validation code:
- unified code for all address validation
- validate sending to burn address in solana
- using name resolution snap for ens names

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/5868
Fixes: https://github.com/MetaMask/MetaMask-planning/issues/5867
Fixes: https://github.com/MetaMask/metamask-mobile/issues/19981

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**
TODO

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Consolidates address/name validation into a single flow using a new name-resolution hook and utilities, replaces legacy EVM/non‑EVM validators, and preinstalls the ENS Resolver Snap.
> 
> - **Send Flow Validation**:
>   - Replace `useEvmToAddressValidation` and `useNonEvmToAddressValidation` with unified logic in `useToAddressValidation` using new utils in `utils/send-address-validations` (hex/Solana burn checks, contract warnings, confusable checks).
>   - Add `useNameValidation` to resolve ENS via Snaps and apply confusable warnings; surface `invalid_address`/could-not-resolve errors.
> - **Snaps Name Resolution**:
>   - Refactor `useSnapNameResolution` to parameterless hook exposing `fetchResolutions(chainId, domain)`; filter snaps by caveats and return `[]` when none available.
> - **UI**:
>   - Adjust `Recipient` page submission/reset behavior; minor dependency cleanup.
> - **Localization**:
>   - Add `send.invalid_address` string; reuse existing messages.
> - **Snaps/Deps**:
>   - Preinstall `@metamask/ens-resolver-snap` and add dependency.
> - **Tests**:
>   - Update and add tests for new hooks and validation utils; adapt existing Snap resolution tests to new API.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 208fe5fcb670a7eb41f50cad18fffad2dd48a9af. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jpuri committed:
Increase pressable area for fiat mode toggling

jpuri committed:
chore: refactor amount validation utility method

jpuri committed:
update

jpuri committed:
Merge branch 'main' into amount_fixes

jpuri committed:
update

jpuri committed:
Adding support to send ERC-1155 tokens

jpuri committed:
Merge branch 'amount_fixes' of https://github.com/MetaMask/metamask-mobile into amount_fixes

jpuri committed:
Merge branch 'main' into amount_fixes

jpuri committed:
Merge branch 'main' into amount_fixes

jpuri committed:
update

jpuri committed:
update

jpuri committed:
Merge branch 'main' into amount_fixes

jpuri committed:
update

jpuri committed:
Merge branch 'amount_fixes' into erc1155_support

jpuri committed:
update

jpuri committed:
Merge branch 'amount_fixes' of https://github.com/MetaMask/metamask-mobile into amount_fixes

jpuri committed:
Merge branch 'main' into amount_fixes

jpuri committed:
Merge branch 'amount_fixes' into erc1155_support

jpuri committed:
update

jpuri committed:
Merge branch 'erc1155_support' of https://github.com/MetaMask/metamask-mobile into erc1155_support

jpuri committed:
Merge branch 'amount_fixes' into erc1155_support

jpuri committed:
update

jpuri committed:
Merge branch 'erc1155_support' of https://github.com/MetaMask/metamask-mobile into erc1155_support

jpuri committed:
update

jpuri committed:
update

jpuri committed:
update

jpuri committed:
update

jpuri committed:
update

jpuri committed:
Refactor recipient validation in new send implementation

jpuri committed:
update

jpuri committed:
update

jpuri committed:
update

jpuri committed:
Merge branch 'main' into recipient_validation

jpuri commented on app/components/Snaps/hooks/useSnapNameResolution.ts:
```diff
@@ -81,8 +81,7 @@ export function useSnapNameResolution({
       );
-
-      if (!cancelled) {
+      if (!cancelled && (!filteredSnaps?.length || responses.length)) {
```
This is to ensure that hook does not return empty result if response is not yet received from snap.

GuillaumeRx replied:
We  should abort earlier if there's no snap available. On top of that there's no way that the hook returns before the promises are all settled 🤔

jpuri replied:
I added early exit here:
```
 if (!filteredSnaps.length || !domain || !chainId) {
      return;
    }
```

The check for `responses.length` is still needed. As I found that promises are resolving with empty result not success or reject, and the hook is able to return empty result for valid domain.

GuillaumeRx replied:
`Promise.allSettled` will always be full if the promises are fulfilled. Please see the docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#return_value

jpuri replied:
I updated the PR to remove `responses.length`, I think there is some other issue with ens snap which is causing it to not return result for valid domains at times.

jpuri committed:
update

jpuri committed:
Merge branch 'recipient_validation' of https://github.com/MetaMask/metamask-mobile into recipient_validation

GuillaumeRx commented on app/components/Snaps/hooks/useSnapNameResolution.ts:
```diff
@@ -58,6 +58,10 @@ export function useSnapNameResolution({
   );
 
   useEffect(() => {
+    if (!filteredSnaps.length || !domain || !chainId) {
```
For sanity

```suggestion
    if (filteredSnaps.length === 0 || !domain || !chainId) {
```

jpuri committed:
Update app/components/Snaps/hooks/useSnapNameResolution.ts

Co-authored-by: Guillaume Roux <guillaumeroux123@gmail.com>

jpuri committed:
Update app/components/Snaps/hooks/useSnapNameResolution.ts

Co-authored-by: Guillaume Roux <guillaumeroux123@gmail.com>

jpuri committed:
update

jpuri committed:
update

jpuri committed:
update

jpuri commented:
I added `responses` to be returned from the hook. As ens snap is repeatedly throwing error:

```
[{"reason": [Error: The snap "npm:@metamask/ens-resolver-snap" has been terminated during execution.], "status": "rejected"}]
```

This is causing UI to show wrong data:


https://github.com/user-attachments/assets/ab3b94bc-3af5-48c7-b88d-55704d48bca8

It works well for solana name resolution snap.

jpuri commented on app/components/Snaps/hooks/useSnapNameResolution.ts:
```diff
@@ -104,5 +111,5 @@ export function useSnapNameResolution({
 
-  return { results, loading };
+  return { responses, results, loading };
 }
```
Currently there is no way to differentiate in:
1. response `fulfilled` with name not resolved
2. response `rejected` for valid name

returning response will enable that

GuillaumeRx committed:
improve snap name resolution by using a callback and cleanup (#20386)

This refactors the snap name lookup hook and does some cleanup around
the logic that uses it.

jpuri committed:
merge

jpuri committed:
Merge branch 'main' into recipient_validation

jpuri committed:
update

jpuri committed:
Merge branch 'recipient_validation' of https://github.com/MetaMask/metamask-mobile into recipient_validation

jpuri committed:
Merge branch 'main' into recipient_validation

jpuri committed:
update

GuillaumeRx commented on app/components/Snaps/hooks/useSnapNameResolution.test.tsx:
Can we add a test for a name service without any available snap ?

jpuri replied:
👍 
I added one more test for chainId with no matching snap.

jpuri committed:
update

jpuri committed:
update

jpuri committed:
Merge branch 'main' into recipient_validation

jpuri committed:
update

jpuri committed:
Merge branch 'recipient_validation' of https://github.com/MetaMask/metamask-mobile into recipient_validation

jpuri committed:
Update app/components/Snaps/hooks/useSnapNameResolution.test.tsx

Co-authored-by: Guillaume Roux <guillaumeroux123@gmail.com>

jpuri committed:
Update

GuillaumeRx reviewed (dismissed):
Snaps changes LGTM

jpuri committed:
update

jpuri committed:
Merge branch 'main' into recipient_validation

jpuri committed:
update

jpuri committed:
Merge branch 'recipient_validation' of https://github.com/MetaMask/metamask-mobile into recipient_validation

jpuri committed:
update

jpuri committed:
Merge branch 'main' into recipient_validation

ngalee8311 commented:
> ## [![Quality Gate Passed](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/checks/QualityGateBadge/qg-passed-20px.png 'Quality Gate Passed')](https://sonarcloud.io/dashboard?id=metamask-mobile&pullRequest=20229) **Quality Gate passed**  
> Issues  
> ![](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/common/passed-16px.png '') [1 New issue](https://sonarcloud.io/project/issues?id=metamask-mobile&pullRequest=20229&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  
> ![](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/common/accepted-16px.png '') [0 Accepted issues](https://sonarcloud.io/project/issues?id=metamask-mobile&pullRequest=20229&issueStatuses=ACCEPTED)
> 
> Measures  
> ![](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/common/passed-16px.png '') [3 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=metamask-mobile&pullRequest=20229&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  
> ![](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/common/passed-16px.png '') [88.4% Coverage on New Code](https://sonarcloud.io/component_measures?id=metamask-mobile&pullRequest=20229&metric=new_coverage&view=list)  
> ![](https://sonarsource.github.io/sonarcloud-github-static-resources/v2/common/passed-16px.png '') [0.0% Duplication on New Code](https://sonarcloud.io/component_measures?id=metamask-mobile&pullRequest=20229&metric=new_duplicated_lines_density&view=list)  
>   
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=metamask-mobile&pullRequest=20229)
> 
>

GuillaumeRx committed:
always return an array in `fetchResolutions`

jpuri committed:
Merge branch 'main' into recipient_validation

FrederikBolding reviewed (approved):
Snaps changes LGTM
